File: zod/src/models/JobOffer.ts
import { z } from 'zod';
import { JobOfferType } from './JobOfferType';
import { ProjectJobOffer } from './ProjectJobOffer';
import { TaskJobOffer } from './TaskJobOffer';
import { CompetitionJobOffer } from './CompetitionJobOffer';

/**
 * 依頼のスキーマ
 */
export const JobOffer = z.discriminatedUnion('type', [
  ProjectJobOffer,
  TaskJobOffer,
  CompetitionJobOffer,
]);

/**
 * 依頼の型
 */
export type JobOffer = z.infer<typeof JobOffer>; 

File: raoh/src/main/java/com/example/raoh/joboffer/data/JobOffer.java
package com.example.raoh.joboffer.data;

import java.time.LocalDate;

/**
 * data 案件 = タイトル AND 詳細 AND 募集期限 AND 案件詳細
 */
public sealed interface JobOffer permits ProjectJobOffer, TaskJobOffer, CompetitionJobOffer {
    String title();
    String description();
    LocalDate offerExpireDate();
}
